### PR TITLE
[YN]Openvino backend

### PR DESCRIPTION
**Pull request by daniil-lyakhov**
### Summary
[PLEASE REMOVE] See [CONTRIBUTING.md's Pull Requests](https://github.com/pytorch/executorch/blob/main/CONTRIBUTING.md#pull-requests) for ExecuTorch PR guidelines.

[PLEASE REMOVE] If this PR closes an issue, please add a `Fixes #<issue-id>` line.

[PLEASE REMOVE] If this PR introduces a fix or feature that should be the upcoming release notes, please add a "Release notes: <area>" label. For a list of available release notes labels, check out [CONTRIBUTING.md's Pull Requests](https://github.com/pytorch/executorch/blob/main/CONTRIBUTING.md#pull-requests).

### Test plan
[PLEASE REMOVE] How did you test this PR? Please write down any manual commands you used and note down tests that you have written if applicable.

### CODE REVIEW

File: examples/openvino/aot/README.md [hunk 1]
```diff
@@ -31,10 +31,17 @@ python aot_openvino_compiler.py --suite <MODEL_SUITE> --model <MODEL_NAME> --inp
   - `[1, 3, 224, 224]` (Zsh users: wrap in quotes)
   - `(1, 3, 224, 224)`
 
+- **`--quantize`** (optional):
+  Enable model quantization: Default is False.
+
+- **`--dataset`** (optional):
+  Path to the calibration dataset. TODO: It is necessary to think in what form to support the dataset. For the experiment, tiny-imagenet is used, which can be downloaded from here http://cs231n.stanford.edu/tiny-imagenet-200.zip and specify the path to it.
```
Comment (alexsu52) [inline examples/openvino/aot/README.md]: ?

File: examples/openvino/aot/aot_openvino_compiler.py [hunk 2]
```diff
@@ -59,22 +146,77 @@ def main(suite: str, model_name: str, input_shape, device: str):
...
+        help="Input shape for the model as a list or tuple (e.g., [1, 3, 224, 224] or (1, 3, 224, 224)).",
+    )
+    parser.add_argument("--quantize", action="store_true", help="Enable model quantization.")
+    parser.add_argument("--dataset", type=str, help="Path to the calibration dataset.")
+    parser.add_argument(
+        "--device",
+        type=str,
+        default="CPU",
+        help="Target device for compiling the model (e.g., CPU, GPU). Default is CPU.",
+    )
```
Comment (alexsu52) [inline examples/openvino/aot/aot_openvino_compiler.py:280]: What do you think about adding a `--qunatization_flow` argument that would have values ​​"pt2e" and "nncf" to specify which flow should be used during quantization?

File: examples/openvino/aot/aot_openvino_compiler.py [hunk 3]
```diff
@@ -4,49 +4,147 @@
...
+from torch.export import export
 from torch.export.exported_program import ExportedProgram
-import argparse
+from torch.fx.passes.graph_drawer import FxGraphDrawer
+from transformers import AutoModel
+
+import nncf
+import nncf.experimental
+import nncf.experimental.torch
+from nncf.experimental.torch.fx.quantization.quantize_pt2e import quantize_pt2e
```
Comment (alexsu52) [inline examples/openvino/aot/aot_openvino_compiler.py]:
```suggestion
import nncf
from nncf.experimental.torch.fx.quantization.quantize_pt2e import quantize_pt2e
```

File: examples/openvino/openvino_build.sh [hunk 4]
```diff
@@ -41,7 +42,7 @@ main() {
           -B"${example_build_dir}" \
           $EXECUTORCH_ROOT/$example_dir
 
-    cmake --build "${example_build_dir}" -j5
+    cmake --build "${example_build_dir}" -j40
```
Comment (alexsu52) [inline examples/openvino/openvino_build.sh]:
```suggestion
    cmake --build "${example_build_dir}" -j5
```

File: examples/openvino/openvino_build.sh [hunk 5]
```diff
@@ -27,12 +27,13 @@ main() {
 
 
     # Build the project
-    cmake --build cmake-openvino-out --target install --config Release -j5
+    cmake --build cmake-openvino-out --target install --config Release -j40
```
Comment (alexsu52) [inline examples/openvino/openvino_build.sh]:
```suggestion
    cmake --build cmake-openvino-out --target install --config Release -j5
```

File: examples/openvino/openvino_build.sh [hunk 6]
```diff
@@ -27,12 +27,13 @@ main() {
...
 
     ## Build example
     local example_dir=examples/openvino
     local example_build_dir="${build_dir}/${example_dir}"
     local cmake_prefix_path="${PWD}/${build_dir}/lib/cmake/ExecuTorch;${PWD}/${build_dir}/third-party/gflags;"
+
```
Comment (alexsu52) [inline examples/openvino/openvino_build.sh]:
```suggestion
```

File: examples/openvino/aot/aot_openvino_compiler.py [hunk 7]
```diff
@@ -59,22 +157,104 @@ def main(suite: str, model_name: str, input_shape, device: str):
...
+        type=str,
+        default="CPU",
+        help="Target device for compiling the model (e.g., CPU, GPU). Default is CPU.",
+    )
 
     args = parser.parse_args()
 
     # Run the main function with parsed arguments
-    main(args.suite, args.model, args.input_shape, args.device)
+    with nncf.torch.disable_patching():
```
Comment (alexsu52) [inline examples/openvino/aot/aot_openvino_compiler.py:285]: Please add comment why we need it.

File: examples/openvino/aot/aot_openvino_compiler.py [hunk 8]
```diff
@@ -59,22 +157,104 @@ def main(suite: str, model_name: str, input_shape, device: str):
...
+        help="Batch size for the validation. Default batch_size == 1."
+        " The dataset length must be evenly divisible by the batch size.",
+    )
+    parser.add_argument("--quantize", action="store_true", help="Enable model quantization.")
+    parser.add_argument(
+        "--validate",
+        action="store_true",
+        help="Enable model validation. --dataset argument is requred for the validation.",
+    )
+    parser.add_argument("--dataset", type=str, help="Path to the calibration dataset.")
```
Comment (alexsu52) [inline examples/openvino/aot/aot_openvino_compiler.py]:
```suggestion
    parser.add_argument("--dataset", type=str, help="Path to the validation dataset.")
```

File: examples/openvino/aot/aot_openvino_compiler.py [hunk 9]
```diff
@@ -4,49 +4,147 @@
...
-    elif not isinstance(input_shape, tuple):
-        raise ValueError("Input shape must be a list or tuple.")
 
+def load_calibration_dataset(dataset_path: str, batch_size: int, suite: str, model: torch.nn.Module, model_name: str):
+    val_dir = f"{dataset_path}/val"
+
+    if suite == "torchvision":
+        transform = torchvision_models.get_model_weights(model_name).DEFAULT.transforms()
+    else:
+        transform = create_transform(**resolve_data_config(model.pretrained_cfg, model=model))
```
Comment (alexsu52) [inline examples/openvino/aot/aot_openvino_compiler.py:60]: Does it work for the huggingface suite?

File: examples/openvino/aot/aot_openvino_compiler.py [hunk 10]
```diff
@@ -4,49 +4,147 @@
...
 
+    if dataset_path:
+        calibration_dataset = load_calibration_dataset(dataset_path, batch_size, suite, model, model_name)
+        input_shape = tuple(next(iter(calibration_dataset))[0].shape)
+        print(f"Input shape retrieved from the model config: {input_shape}")
+    # Ensure input_shape is a tuple
+    elif isinstance(input_shape, list):
+        input_shape = tuple(input_shape)
+    else:
+        msg = "Input shape must be a list or tuple."
```
Comment (alexsu52) [inline examples/openvino/aot/aot_openvino_compiler.py:181]: It looks like if input_shape is tuple I will be here.

File: examples/openvino/aot/aot_openvino_compiler.py [hunk 11]
```diff
@@ -4,49 +4,147 @@
...
+        raise ValueError(msg)
     # Provide input
-    example_args = (torch.randn(*input_shape), )
+    example_args = (torch.randn(*input_shape),)
 
-    # Export to aten dialect using torch.export
+    # Export the model to the aten dialect
     aten_dialect: ExportedProgram = export(model, example_args)
 
+    if quantize:
```
Comment (alexsu52) [inline examples/openvino/aot/aot_openvino_compiler.py:189]: Could you implement quantization flow in the separate function?

File: examples/openvino/aot/aot_openvino_compiler.py [hunk 12]
```diff
@@ -59,22 +157,104 @@ def main(suite: str, model_name: str, input_shape, device: str):
...
 
     # Serialize and save it to a file
-    with open(f"{model_name}.pte", "wb") as file:
+    model_file_name = f"{model_name}_{'int8' if quantize else 'fp32'}.pte"
+    with open(model_file_name, "wb") as file:
         exec_prog.write_to_file(file)
-    print(f"Model exported and saved as {model_name}.pte on {device}.")
+    print(f"Model exported and saved as {model_file_name} on {device}.")
+
+    if validate:
```
Comment (alexsu52) [inline examples/openvino/aot/aot_openvino_compiler.py:221]: Could you implement validation in the separate function?

File: backends/openvino/quantizer/quantizer.py [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+
+import nncf
+import nncf.common.quantization as q
+import nncf.experimental.torch.fx as nncf_fx
+import nncf.parameters as p
+import nncf.quantization.advanced_parameters as advanced_p
+from nncf.common.graph.graph import NNCFGraph
+
+QUANT_ANNOTATION_KEY = "quantization_annotation"
+
```
Comment (alexsu52) [inline backends/openvino/quantizer/quantizer.py:27]: I suggest to introduce the following class:

```
class QuantizationMode(StrEnum):
    """
    Defines special quantization modes.
 
    - INT8_SYM: INT8 symmetric quantization for both activations and weights.
    - INT8_MIXED: INT8 asymmetric quantization for activations, symmetric for weights.
    - INT8_TRANSFORMER: Optimized INT8 quantization for transformer-based models
    """
 
    INT8_SYM = "int8_sym"
    INT8_MIXED = "int8_mixed"
    INT8_TRANSFORMER = "int8_transformer"
```

File: backends/openvino/quantizer/quantizer.py [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+        mode: Optional[p.QuantizationMode] = None,
+        preset: Optional[q.structs.QuantizationPreset] = None,
+        target_device: p.TargetDevice = p.TargetDevice.ANY,
+        transformer_model: bool = False,
+        ignored_scope: Optional[nncf.IgnoredScope] = None,
+        overflow_fix: Optional[advanced_p.OverflowFix] = None,
+        quantize_outputs: bool = False,
+        activations_quantization_params: Optional[advanced_p.QuantizationParameters] = None,
+        weights_quantization_params: Optional[advanced_p.QuantizationParameters] = None,
+    ):
```
Comment (alexsu52) [inline backends/openvino/quantizer/quantizer.py]: What do you think?

```suggestion
    def __init__(
        self,
        *,
        mode: QuantizationMode = QuantizationMode.INT8_MIXED,
        ignored_scope: Optional[nncf.IgnoredScope] = None,
        **kwargs
    ):
```

File: examples/openvino/aot/README.md (hunk 1 above)
Comment (daniil-lyakhov) [reply examples/openvino/aot/README.md]: Fixed

File: examples/openvino/aot/aot_openvino_compiler.py (hunk 2 above)
Comment (daniil-lyakhov) [reply examples/openvino/aot/aot_openvino_compiler.py:280]: Done, please check

File: examples/openvino/aot/aot_openvino_compiler.py (hunk 7 above)
Comment (daniil-lyakhov) [reply examples/openvino/aot/aot_openvino_compiler.py:285]: Done

File: examples/openvino/aot/aot_openvino_compiler.py (hunk 9 above)
Comment (daniil-lyakhov) [reply examples/openvino/aot/aot_openvino_compiler.py:60]: No, an exception is added

File: examples/openvino/aot/aot_openvino_compiler.py (hunk 10 above)
Comment (daniil-lyakhov) [reply examples/openvino/aot/aot_openvino_compiler.py:181]: Done

File: examples/openvino/aot/aot_openvino_compiler.py (hunk 11 above)
Comment (daniil-lyakhov) [reply examples/openvino/aot/aot_openvino_compiler.py:189]: Done

File: examples/openvino/aot/aot_openvino_compiler.py (hunk 12 above)
Comment (daniil-lyakhov) [reply examples/openvino/aot/aot_openvino_compiler.py:221]: Done

File: backends/openvino/quantizer/quantizer.py (hunk 14 above)
Comment (daniil-lyakhov) [reply backends/openvino/quantizer/quantizer.py]: I like the constructor parameters subset, refactored

File: backends/openvino/quantizer/quantizer.py (hunk 13 above)
Comment (daniil-lyakhov) [reply backends/openvino/quantizer/quantizer.py:27]: Done

File: backends/openvino/quantizer/quantizer.py (hunk 14 above)
Comment (daniil-lyakhov) [reply backends/openvino/quantizer/quantizer.py]: Also, I set default QuantizationMode == INT8_SYM to alight with the default MinMax parameters (which were default parameters before) https://github.com/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/min_max/algorithm.py#L211-L215

File: backends/openvino/quantizer/quantizer.py [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+    """
+    Implementation of the Torch AO quantizer which annotates models with quantization annotations
+    optimally for the inference via OpenVINO.
+    """
+
+    def __init__(
+        self,
+        *,
+        mode: Optional[QuantizationMode] = QuantizationMode.INT8_SYM,
+        ignored_scope: Optional[nncf.IgnoredScope] = None,
```
Comment (alexsu52) [inline backends/openvino/quantizer/quantizer.py]: Ignored scope is model specific information. What do you think about introducing additional method to set ignored scope?
```
def set_ignored_scope(names: Optional[List[str]] = None, patterns: Optional[List[str]] = None, types: Optional[List[str]] = None, subgraphs: Optional[List[Tuple[List[str], List[str]]]] = None, validate: bool = True)
```

Comment (daniil-lyakhov) [reply backends/openvino/quantizer/quantizer.py]: Done
